js-runtime/index.js

```diff
@@ -142,59 +166,469 @@ module.exports = {
   boolstr:  a => a.toString(),
 
   // Arithmetic opcodes
-  addi8:   (a, b) => a + b,
-  addi16:  (a, b) => a + b,
-  addi32:  (a, b) => a + b,
-  addi64:  (a, b) => a + b,
-  addf32:  (a, b) => a + b,
-  addf64:  (a, b) => a + b,
-
-  subi8:   (a, b) => a - b,
-  subi16:  (a, b) => a - b,
-  subi32:  (a, b) => a - b,
-  subi64:  (a, b) => a - b,
-  subf32:  (a, b) => a - b,
-  subf64:  (a, b) => a - b,
+  addi8:   (ra, rb) => {
+    if (!ra[0]) return ra
+    if (!rb[0]) return rb
+    const a = ra[1]
+    const b = rb[1]
+    if (a === 0 || b === 0) return [1, a + b]
+    if (a < 0 && b > 0) return [1, a + b]
+    if (a > 0 && b < 0) return [1, a + b]
+    if (a > 0 && b > 0) {
+      if (a > INT8MAX - b) return [0, 'overflow']
+      return [1, a + b]
+    }
+    if (a < 0 && b < 0) {
+      if (a < INT8MIN - b) return [0, 'underflow']
+      return [1, a + b]
+    }
+  },
+  addi16:  (ra, rb) => {
+    if (!ra[0]) return ra
+    if (!rb[0]) return rb
+    const a = ra[1]
+    const b = rb[1]
+    if (a === 0 || b === 0) return [1, a + b]
+    if (a < 0 && b > 0) return [1, a + b]
+    if (a > 0 && b < 0) return [1, a + b]
+    if (a > 0 && b > 0) {
+      if (a > INT16MAX - b) return [0, 'overflow']
+      return [1, a + b]
+    }
+    if (a < 0 && b < 0) {
+      if (a < INT16MIN - b) return [0, 'underflow']
+      return [1, a + b]
+    }
+  },
+  addi32:  (ra, rb) => {
+    if (!ra[0]) return ra
+    if (!rb[0]) return rb
+    const a = ra[1]
+    const b = rb[1]
+    if (a === 0 || b === 0) return [1, a + b]
+    if (a < 0 && b > 0) return [1, a + b]
+    if (a > 0 && b < 0) return [1, a + b]
+    if (a > 0 && b > 0) {
+      if (a > INT32MAX - b) return [0, 'overflow']
+      return [1, a + b]
+    }
+    if (a < 0 && b < 0) {
+      if (a < INT32MIN - b) return [0, 'underflow']
+      return [1, a + b]
+    }
+  },
+  addi64:  (ra, rb) => {
+    if (!ra[0]) return ra
+    if (!rb[0]) return rb
+    const a = ra[1]
+    const b = rb[1]
+    if (a === 0n || b === 0n) return [1, a + b]
+    if (a < 0n && b > 0n) return [1, a + b]
+    if (a > 0n && b < 0n) return [1, a + b]
+    if (a > 0n && b > 0n) {
+      if (a > INT64MAX - b) return [0, 'overflow']
+      return [1, a + b]
+    }
+    if (a < 0n && b < 0n) {
+      if (a < INT64MIN - b) return [0, 'underflow']
+      return [1, a + b]
+    }
+  },
+  addf32:  (ra, rb) => {
+    if (!ra[0]) return ra
+    if (!rb[0]) return rb
+    const a = ra[1]
+    const b = rb[1]
+    const out = a + b
+    if (out === Number.POSITIVE_INFINITY) return [0, 'overflow']
+    if (out === Number.NEGATIVE_INFINITY) return [0, 'underflow']
+    return [1, out]
+  },
```

Comment:
While it's kinda dumb for floating point numbers, since they can't actually underflow/overflow/etc from a CPU hardware perspective, doing this keeps it consistent with the integers and forces you to deal with error situations where not even 64-bit floats could represent your number.

Reply:
And the proposed "saturating math" operators restore the "normal" floating point behavior and give integers a similar behavior.